### PR TITLE
feat: support npm resolution for file imports

### LINKED ISSUE #4098: 🚀 Feature Request: resolve npm exports for wasm (etc.) imports
## Problem

Today, if you use an npm module that ships a wasm package such as https://github.com/ssssota/svg2png-wasm, the ergonomics around importing it are less than ideal.

You essentially have to import it like so:
```typescript
import wasm from '../../node_modules/svg2png-wasm/svg2png_wasm_bg.wasm'
```

Or, you have a build script that copies it out of `node_modules` into somewhere else in your app, and then import it from there.

Attempts to import it using the more familiar and ergonomic syntax you'll find in the JS ecosystem such as this:
```typescript
import wasm from 'svg2png-wasm/svg2png_wasm_bg.wasm'
```

results in an error like this, relative to where it was imported:
```
✘ [ERROR] ENOENT: no such file or directory, open 'xxx\src\libs\svg2png-wasm\svg2png_wasm_bg.wasm' [plugin wrangler-module-collector]

    src/libs/qrcode.ts:9:17:
      9 │ import wasm from 'svg2png-wasm/svg2png_wasm_bg.wasm';
```

This package in particular makes this wasm file available in its export map: https://github.com/ssssota/svg2png-wasm/blob/7e89e869b501df458d5544d0eaea067d0fdc8b3d/package.json#L17, so this should "just work".

## Possible Fixes

It seems like I'm hitting `wrangler-module-collector`, and this codepath: https://github.com/cloudflare/workers-sdk/blob/80f78dad4652cb9c6807c5144c4c32324f0c15e6/packages/wrangler/src/module-collection.ts#L275

Adding some resolution logic for npm paths such as via https://www.npmjs.com/package/resolve.exports could really improve the DX here.

### PR DESCRIPTION
**Pull request by Cherry**
Fixes #4098.

**What this PR solves / how to test:**

Previously when importing static files (such as wasm) from an npm package, you had to path directly to it, such as :
```javascript
import wasm from '../../node_modules/svg2png-wasm/svg2png_wasm_bg.wasm'
```
While this works file, the DX isn't ideal, so this PR adds proper package resolution for these statements, allowing you to do:
```javascript
import wasm from 'svg2png-wasm/svg2png_wasm_bg.wasm'
```

This uses [`resolve.exports`](https://www.npmjs.com/package/resolve.exports) to look at and analyse the package's `exports` field, and resolve to the applicable file. This does add an additional dependency, but this is a very common one used by so much build tooling including `vite`, `jest`, `svelte`, and more.

If for any reason this fails, it falls back to the previous behaviour and will just do a path reference, and then end up throwing an `ENOENT` as per previous functionality.

Notes:

- My detection for "is this possibly an npm package" is very dumb right now, checking for presence of `/` and not starting with `.`. It seems to work remarkebly well in practice across the many projects I've tested this against, but I'm happy to improve this if someone has a cleaner idea.
- While I've added tests for the npm package name extraction to ensure it handles scoped packages, I've not added any specific tests for handling of the imports. To test that, it'd need a fixture with a real `node_modules` folder, and dummy package, which I don't think there's any prior art for in the repo today. Happy to add these if someone can guide me on the best way to approach this.

**Author has included the following, where applicable:**

- [x] Tests
- [x] Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))

**Reviewer is to perform the following, as applicable:**

- Checked for inclusion of relevant tests
- Checked for inclusion of a relevant changeset
- Checked for creation of associated docs updates
- Manually pulled down the changes and spot-tested

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

### CODE REVIEW

File: packages/wrangler/src/deployment-bundle/module-collection.ts [hunk 1]
```diff
@@ -251,6 +252,42 @@ export function createModuleCollector(props: {
 								// it to `esbuild` to bundle it.
 								if (isJavaScriptModuleRule(rule)) return;
 
+								// Check if this file is possibly from an npm package
+								// and if so, validate the import against the package.json exports
+								// and resolve the file path to the correct file.
+								if (args.path.includes("/") && !args.path.startsWith("./")) {
```
Comment (Skye-31) [inline packages/wrangler/src/deployment-bundle/module-collection.ts:258]: This should cover more cases https://nodejs.org/api/path.html#pathisabsolutepath
```suggestion
								if (args.path.includes("/") && !path.isAbsolute(args.path)) {
```

Comment (Skye-31) [reply packages/wrangler/src/deployment-bundle/module-collection.ts:258]: Oh I misread it, nevermind

**Comment by petebacondarwin** [comment]
Can this be done more simply using the node.js `require.resolve()` function? For example something like:

```ts
let filePath = path.join(args.resolveDir, args.path);
if (!path.isAbsolute(args.path) && !args.path.startsWith(".")) {
	try {
	filePath = require.resolve(args.path)
} catch {
	// Do nothing
}
```

Note that your current checks for package like paths would break on Windows.

**Comment by Cherry** [comment]
> Can this be done more simply using the node.js `require.resolve()` function?

Unfoortunately not from my testing. Running `require.resolve` with `args.path` as `svg2png-wasm/svg2png_wasm_bg.wasm` results in `Error: Cannot find module 'svg2png-wasm/svg2png_wasm_bg.wasm'` with code `MODULE_NOT_FOUND `. I don't know exactly why this is, but I'd suspect it's something to do with CJS vs ESM interop, and newer exports mapping.

> Note that your current checks for package like paths would break on Windows.

Can you expand here? My checks are purely checking for an included `/`, and not starting with `.`. 

If I try and import like `svg2png-wasm\\svg2png_wasm_bg.wasm`, this plugin doesn't even run, and I get a generic esbuild error: `No loader is configured for ".wasm" files: ../../../../example-wasm/node_modules/svg2png-wasm/svg2png_wasm_bg.wasm`. I suspect this is because this plugin only runs on the specified filter, which looks something like this:
```
globs: ["**/*.wasm", "**/*.wasm?module"]
```

Supporting more windows orientated backslashes seems like a general issue here, and not really in the scope of this PR if I'm understanding correctly. It's also extremely uncommon, even on Windows from my experience, to import using backslashes in the path, especially with npm modules. This would also explain why no one has reported it as a bug.

File: packages/wrangler/src/deployment-bundle/module-collection.ts [hunk 2]
```diff
@@ -64,6 +65,23 @@ export const noopModuleCollector: ModuleCollector = {
...
 };
 
+// Extracts a package name from a string that may be a file path
+// or a package name. Returns null if the string is not a valid
+// Handles `wrangler`, `wrangler/example`, `wrangler/example.wasm`,
+// `@cloudflare/wrangler`, `@cloudflare/wrangler/example`, etc.
+export function extractPackageName(packagePath: string) {
+	if (packagePath.startsWith(".")) return null;
+
+	const match = packagePath.match(/^(@[^/]+\/)?([^/]+)/);
```
Comment (petebacondarwin) [inline packages/wrangler/src/deployment-bundle/module-collection.ts:75]: This would match `c:\foo\bar`.

**Comment by petebacondarwin** [comment]
> Supporting more windows orientated backslashes seems like a general issue here, and not really in the scope of this PR if I'm understanding correctly. It's also extremely uncommon, even on Windows from my experience, to import using backslashes in the path, especially with npm modules. This would also explain why no one has reported it as a bug.

Yeah that is a good spot. The rules themselves wouldn't match in the first place.

File: packages/wrangler/src/deployment-bundle/module-collection.ts (hunk 2 above)
Comment (Cherry) [reply packages/wrangler/src/deployment-bundle/module-collection.ts:75]: Do you think I should explicitly just disallow `\` in the regex for now? Until this plugin in general handles backslashes for windows paths?

Comment (Cherry) [reply packages/wrangler/src/deployment-bundle/module-collection.ts:75]: Awesome, thanks! Let me know if you need anything else on my end before merge!

**Comment by Cherry** [comment]
All of the E2E test failures look unrelated, failing due to `AssertionError: Please provide a CLOUDFLARE_ACCOUNT_ID as an environment variable`

File: .changeset/long-starfishes-mate.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+---
+"wrangler": patch
```
Comment (penalosa) [inline .changeset/long-starfishes-mate.md:2]:
```suggestion
"wrangler": minor
```

File: packages/wrangler/src/deployment-bundle/module-collection.ts [hunk 4]
```diff
@@ -251,6 +269,51 @@ export function createModuleCollector(props: {
...
+									// and then finally override filePath if we find a match.
+									try {
+										const packageJson = JSON.parse(
+											await readFile(packageJsonPath, "utf8")
+										);
+										const testResolved = resolveExports(
+											packageJson,
+											args.path.replace(`${packageName}/`, ""),
+											{
+												conditions: ["workerd", "worker", "browser"],
```
Comment (penalosa) [inline packages/wrangler/src/deployment-bundle/module-collection.ts:300]: Can you pull this out into a constant for sharing between here and https://github.com/cloudflare/workers-sdk/blob/f4ad634af86c49ade427af23e3853c656e30250a/packages/wrangler/src/deployment-bundle/bundle.ts#L313?

**Review by penalosa — COMMENTED** [review]
This is looking great! I would like to see some tests of this behaviour: perhaps similarly to the `worker-app` example, which imports from another fixture, you could write a fixture which imports a wasm file from another fixture?

File: .changeset/long-starfishes-mate.md (hunk 3 above)
Comment (Cherry) [reply .changeset/long-starfishes-mate.md]: Thanks, done

File: packages/wrangler/src/deployment-bundle/module-collection.ts (hunk 4 above)
Comment (Cherry) [reply packages/wrangler/src/deployment-bundle/module-collection.ts]: Done, lemme know if you'd prefer this in another file

**Comment by Cherry** [comment]
Thanks! I've added two test fixtures now - one just exporting a simple `multiply.wasm` file (stolen from another fixture), and then the other importing it. In doing so, I found some small issues with how I was using `resolveDir` vs `process.cwd()` so got that all fixed up.

Let me know if there's any further changes needed here.

**Review by Skye-31 — APPROVED** [review]

**Comment by Cherry** [comment]
The failing tests seem to be in `external-service-bindings-app:test:ci`, and unrelated to this PR.

**Comment by Cherry** [comment]
Nevermind, something in `service-bindings-app` is still flaking and throwing a v8 error. Completely unrelated to this PR, but preventing it from going green.

**Pull request merged**